app/controllers/votes_controller.rb

```diff
@@ -0,0 +1,41 @@
+class VotesController < ApplicationController
+  def upvote
+    found = Votes.find_by(user: params[:user], api: params[:api_id])
+
+    if found == nil
+      Votes.create(
+        user: User.find(params[:user]),
+        api: Api.find(params[:api_id]),
+        user_vote: true)
+      render json: Votes.calculate_votes(params[:api_id])
+      return
+    end
+    if found.user_vote == false
+      found.user_vote = true
+      found.save
+      render json: Votes.calculate_votes(params[:api_id])
+      return
+    end
+    render json: Votes.calculate_votes(params[:api_id])
+  end
+
+  def downvote
+    found = Votes.find_by(user: params[:user], api: params[:api_id])
+
+    if found == nil
+      Votes.create(
+        user: User.find(params[:user]),
+        api: Api.find(params[:api_id]),
+        user_vote: false)
+      render json: Votes.calculate_votes(params[:api_id])
+      return
+    end
+    if found.user_vote == true
+      found.user_vote = false
+      found.save
+      render json: Votes.calculate_votes(params[:api_id])
+      return
+    end
+    render json: Votes.calculate_votes(params[:api_id])
+  end
+end
```

Comment:
After reading all the logic in this controller, there definitely need to be some methods on the `Vote` model that do some of this work. The controller methods are definitely longer than they ought to be in an ideal application.